pageserver/benches/bench_walredo.rs

```diff
@@ -1,160 +1,151 @@
+//! short/2                 time:   [27.216 µs 27.372 µs 27.557 µs]
+//! short/4                 time:   [44.398 µs 45.858 µs 47.178 µs]
+//! short/8                 time:   [81.236 µs 83.332 µs 85.419 µs]
+//! short/16                time:   [138.29 µs 139.76 µs 141.24 µs]
+//! short/32                time:   [149.67 µs 150.44 µs 151.30 µs]
+//! short/64                time:   [155.31 µs 155.90 µs 156.59 µs]
+//! short/128               time:   [156.33 µs 156.85 µs 157.44 µs]
+//! medium/1                time:   [105.47 µs 105.87 µs 106.36 µs]
+//! medium/2                time:   [157.09 µs 157.68 µs 158.40 µs]
+//! medium/4                time:   [293.69 µs 306.80 µs 318.11 µs]
+//! medium/8                time:   [594.88 µs 614.05 µs 633.18 µs]
+//! medium/16               time:   [848.28 µs 853.06 µs 858.68 µs]
+//! medium/32               time:   [916.31 µs 920.97 µs 926.27 µs]
+//! medium/64               time:   [939.99 µs 945.49 µs 951.30 µs]
+//! medium/128              time:   [918.89 µs 928.75 µs 938.41 µs]
+//! ```
 
 use bytes::{Buf, Bytes};
-use pageserver::{
-    config::PageServerConf, repository::Key, walrecord::NeonWalRecord, walredo::PostgresRedoManager,
+use criterion::{BenchmarkId, Criterion};
+use pageserver::{config::PageServerConf, walrecord::NeonWalRecord, walredo::PostgresRedoManager};
+use pageserver_api::{key::Key, shard::TenantShardId};
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
 };
-use pageserver_api::shard::TenantShardId;
-use tokio::task::JoinSet;
+use tokio::{sync::Barrier, task::JoinSet};
 use utils::{id::TenantId, lsn::Lsn};
 
-use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
+fn bench(c: &mut Criterion) {
+    {
+        let nclients = [1, 2, 4, 8, 16, 32, 64, 128];
+        for nclients in nclients {
+            let mut group = c.benchmark_group("short");
+            group.bench_with_input(
+                BenchmarkId::from_parameter(nclients),
+                &nclients,
+                |b, nclients| {
+                    let redo_work = Arc::new(Request::short_input());
+                    b.iter_custom(|iters| bench_impl(Arc::clone(&redo_work), iters, *nclients));
+                },
+            );
+        }
+    }
 
-fn redo_scenarios(c: &mut Criterion) {
-    // logging should be enabled when adding more inputs, since walredo will only report malformed
-    // input to the stderr.
-    // utils::logging::init(utils::logging::LogFormat::Plain).unwrap();
+    {
+        let nclients = [1, 2, 4, 8, 16, 32, 64, 128];
+        for nclients in nclients {
+            let mut group = c.benchmark_group("medium");
+            group.bench_with_input(
+                BenchmarkId::from_parameter(nclients),
+                &nclients,
+                |b, nclients| {
+                    let redo_work = Arc::new(Request::medium_input());
+                    b.iter_custom(|iters| bench_impl(Arc::clone(&redo_work), iters, *nclients));
```

Comment:
So by using iter_custom, one avoids timing the runtime setup?

Reply:
Yes, see the description at the top of the file.

Reply:
Surprising place to add the docs, but looks ok.